### ISSUE TITLE
The driver doesn't send requests to all nodes

### ISSUE THREAD

**Issue by Kulezi** [scylladb/scylla-rust-driver#487]
![image](https://user-images.githubusercontent.com/20567521/181788433-d6969579-3826-4035-b136-9ba7ea6b8106.png)
In a setup with 3 nodes RF=1, driver provided only 1 host, (yellow line), provided host doesn't receive any queries.
https://github.com/Kulezi/benchtabs/blob/main/scylla-rust-driver/src/main.rs

This happens on main (https://github.com/scylladb/scylla-rust-driver/commit/90044469948bc31fcc7c794d88eb245e2607a68a), but doesn't on release 0.4.7

cc: @wprzytula

**Comment by psarna** [comment]
hm, there weren't that many commits between 0.4.7 and master and none look relevant at first glance.
```
9004446 (HEAD, origin/main, origin/HEAD) Merge pull request #486 from Gor027/maybe_unset_clone_impl
38d909d Add Clone/Copy to MaybeUnset and Clone to SessionBuilder
e093bbd Merge pull request #471 from wprzytula/address-translator
e66c6d8 Optional address translation added
014c83a Merge pull request #479 from wprzytula/tests-in-docker
46fe1dc Added scripts for running tests in docker.
f68ba27 Merge pull request #473 from cvybhu/unprepared
8fd3b41 caching_session: execute: Take ValueList by value (#474)
a5e492e caching_session: Remove repreparing logic from CachingSession
26a8bce session_test: Test repreparation in CachingSession::execute()
fc9fff1 session_test: Test repreparation in Session::batch()
94ff24e session_test: Test repreparation in Session::execute()
d3d3dc8 connection: Add logs when repreparing a PreparedStatement
3359c9c Remove sleeping to await schema agreement from tests (#466)
9d402a9 examples/allocations: don't spawn new tokio tasks for each request (#461)
```

Did you observe this behavior more than once, does it reproduce every time? If it does, I'll try to bisect this issue next week and find the offending change, and then post tracing results here.

**Comment by Kulezi** [comment]
Yes, it happened on each run of the bencharks we ran with @avelanarius , also please note that the provided node address was local

**Comment by cvybhu** [comment]
I ran into something similar.
I started scylla in the background using:
```bash
podman run --rm -it -p 9042:9042 scylladb/scylla
```

And then tried to run the basic example, but the connection failed:
```bash
$ cargo run --example basic
    Finished dev [unoptimized + debuginfo] target(s) in 0.06s
     Running `target/debug/examples/basic`
Connecting to 127.0.0.1:9042 ...
Error: IO Error: No connections in the pool; last connection failed with: IO Error: No route to host (os error 113)

```

Reverting e66c6d8f20817a119d5f813eab191d286a110f1d helped, so it  might be an issue with the new address translation, although I haven't done any bisecting yet.

**Comment by cvybhu** [comment]
Yup my issue with connecting bisected to e66c6d8f20817a119d5f813eab191d286a110f1d, maybe one node is unable to connect because of #471 as well?

**Comment by avelanarius** [comment]
I think it’ll be the best for @wprzytula to try to understand and fix the problem - it might be just a trivial regression after #471.

**Comment by piodul** [comment]
> I ran into something similar. I started scylla in the background using:
> 
> ```shell
> podman run --rm -it -p 9042:9042 scylladb/scylla
> ```
> 
> And then tried to run the basic example, but the connection failed:
> 
> ```shell
> $ cargo run --example basic
>     Finished dev [unoptimized + debuginfo] target(s) in 0.06s
>      Running `target/debug/examples/basic`
> Connecting to 127.0.0.1:9042 ...
> Error: IO Error: No connections in the pool; last connection failed with: IO Error: No route to host (os error 113)
> ```
> 
> Reverting [e66c6d8](https://github.com/scylladb/scylla-rust-driver/commit/e66c6d8f20817a119d5f813eab191d286a110f1d) helped, so it might be an issue with the new address translation, although I haven't done any bisecting yet.

I confirm, I get the same error when I run scylla using `podman`. However, the example works fine when running scylla with `docker`. Perhaps the internal IPs are not reachable for podman containers, just like they aren't for docker containers running on non-Linux systems. We used to have a workaround for this issue as described [here](https://github.com/scylladb/scylla-rust-driver/issues/476) but we removed it as it didn't work well with address translation.

This might explain your issue with `podman`, but I don't think this would explain the problem described in the issue.

**Comment by psarna** [comment]
@avelanarius @wprzytula let's try to understand and fix, and let me know by the end of the day if you've pinpointed the issue. If nothing's found today, then I'll just temporarily revert this series so that the master branch is not affected for too long, and we can re-merge it once we figure out what the regression is.

**Comment by wprzytula** [comment]
I've found the cause of the issue: the `rpc_address` of the node provided is set to `0.0.0.0` in `system.local` table. It is subject to further discussion how we should deal with such (mis?)configuration.

**Comment by psarna** [comment]
Ref, this is how gocql prioritizes addresses from system.peers when trying to connect: maybe we should also mimic this behavior to be more resilient:
https://github.com/scylladb/gocql/blob/e6fdf892f469b5ad7b9ee4359015d815363422d5/host_source.go#L165-L179

**Comment by wprzytula** [comment]
The issue should be solved by #496, because I've combined address translation with a special case for contact points:
1. Instead of using `rpc_address` from `system.local`, we stick to `connection_address` - the one given as `known_host` upon `Session` construction.
2. Initially-provided addresses (contact points) are retained in driver structs and never undergo address translation.

This prevents the problem described in this issue.

**Comment by psarna** [comment]
@wprzytula if #496 contains the fix, please split it out of the big series so that it's a separate small PR, which we can merge faster and mark that it fixes this issue

**Comment by wprzytula** [comment]
@piodul Address translation #509 contained fix for this, so let's close the issue.

**Comment by piodul** [comment]
I confirm that, with the new implementation, requests are routed to all nodes and the control connection node is not skipped (at least when address translation is turned off). Closing.
